evm/src/executor/fork/cache.rs

```diff
@@ -0,0 +1,234 @@
+//! Cache related abstraction
+use ethers::types::{Address, H256, U256};
+use parking_lot::RwLock;
+use revm::AccountInfo;
+use serde::{ser::SerializeMap, Deserialize, Deserializer, Serialize, Serializer};
+use std::{collections::BTreeMap, fs, io::BufWriter, path::PathBuf, sync::Arc};
+use tracing::{trace, trace_span, warn};
+use tracing_error::InstrumentResult;
+
+pub type StorageInfo = BTreeMap<U256, U256>;
+
+/// A shareable Block database
+#[derive(Clone)]
+pub struct BlockchainDb {
+    /// Contains all the data
+    db: Arc<MemDb>,
+    /// metadata of the current config
+    meta: Arc<RwLock<BlockchainDbMeta>>,
+    /// the cache that can be flushed
+    cache: Arc<dyn BlockCacheDB>,
+}
+
+impl BlockchainDb {
+    /// Creates a new instance of the [BlockchainDb]
+    ///
+    /// if a `cache_path` is provided it attempts to load a previously stored [JsonBlockCacheData]
+    /// and will try to use the cached entries it holds.
+    ///
+    /// This will return a new and empty [MemDb] if
+    ///   - `cache_path` is `None`
+    ///   - the file the `cache_path` points to, does not exist
+    ///   - the file contains malformed data, or if it couldn't be read
+    ///   - the provided `meta` differs from [BlockchainDbMeta] that's stored on disk
+    pub fn new(meta: BlockchainDbMeta, cache_path: Option<PathBuf>) -> Self {
+        // read cache and check if metadata matches
+        let cache = cache_path
+            .and_then(|p| {
+                JsonBlockCacheDB::load(p).ok().filter(|cache| {
+                    if meta != *cache.meta().read() {
+                        warn!(target:"cache", "non-matching block metadata");
+                        false
+                    } else {
+                        true
+                    }
+                })
+            })
+            .unwrap_or_else(|| JsonBlockCacheDB::transient(Arc::new(RwLock::new(meta))));
```

Comment:
Should we log a warning or error here, so it doesn't fail silently?

Reply:
is logged in `load`